README.md
  
## Getting Started  
  
To use this plugin, add `expandable_text_widget` as a [dependency in your pubspec.yaml file](https://pub.dev/packages/expandable_text_widget).  

### Demo 


# <img src="https://github.com/SunnatilloQuantic/Expandeble-Text-Widget/blob/main/assets/demo.jpg"  width="160" height="320" >


### Example  
  
```import 'package:sunnatillo_quanticuz/expandable_text_widget.dart';
   import 'package:flutter/material.dart';
   import 'package:flutter_svg/flutter_svg.dart';
   
   void main() {
     runApp(MyApp());
   }
   
   class MyApp extends StatelessWidget {
     @override
     Widget build(BuildContext context) {
       return MaterialApp(
         title: 'Expandable Text Widget Demo',
         theme: ThemeData(
           primarySwatch: Colors.blue,
         ),
         home: ExpandableDemo(),
         debugShowCheckedModeBanner: false,
       );
     }
   }
   
   class ExpandableDemo extends StatefulWidget {
     @override
     _ExpandableDemoState createState() => _ExpandableDemoState();
   }
   
   class _ExpandableDemoState extends State<ExpandableDemo> {
     @override
     Widget build(BuildContext context) {
       return Scaffold(
         appBar: AppBar(
           title: Text("Expandable Text Widget Demo"),
         ),
         body: SingleChildScrollView(
           child: Container(
             child: Column(
               mainAxisAlignment: MainAxisAlignment.start,
               crossAxisAlignment: CrossAxisAlignment.stretch,
               children: <Widget>[
                 ExpandableTextWidget(
                   backgroundColor: Colors.white,
                   elevation: 2.0,
                   shape: RoundedRectangleBorder(
                     borderRadius: BorderRadius.all(Radius.circular(8.0)),
                   ),
                   title: Text(
                     "Sarlavha",
                     style: TextStyle(color: Colors.grey[800], fontSize: 24),
                   ),
                   imageChild: Image.asset(
                     "assets/logo.png",
                     fit: BoxFit.scaleDown,
                   ),
                   textStyle: TextStyle(color: Colors.black, fontSize: 16),
                   text:
                       "The color can \n be set with the splashColor property. \nThe splash size is \ndependent on the size of the child widget passed in - which is constrained by the minRadius and maxRadius parameters.",
                   downIcon: SvgPicture.asset(
                     "assets/down.svg",
                     color: Colors.black,
                     fit: BoxFit.scaleDown,
                   ),
                   upIcon: SvgPicture.asset(
                     "assets/up.svg",
                     color: Colors.black,
                     fit: BoxFit.scaleDown,
                   ),
                 ),
                 SizedBox(height: 20,),
                 ExpandableTextWidget(
                   backgroundColor: Colors.white,
                   elevation: 2.0,
                   shape: RoundedRectangleBorder(
                     borderRadius: BorderRadius.all(Radius.circular(8.0)),
                   ),
                   title: Text(
                     "Sarlavha",
                     style: TextStyle(color: Colors.grey[800], fontSize: 24),
                   ),
                   textStyle: TextStyle(color: Colors.black, fontSize: 16),
                   text:
                   "The color can \n be set with the splashColor property. \nThe splash size is \ndependent on the size of the child widget passed in - which is constrained by the minRadius and maxRadius parameters.",
                   downIcon: Text("show"),
                   upIcon: Text("hide"),
                 ),
               ],
             ),
           ),
         ),
       );
     }
   }
```


CHANGELOG.md
## [0.0.1] - 2020/10/2
* initial release.

## [0.0.2] - 2020/10/2
* update readme.

## [0.0.3] - 2020/10/2
* change icon.

## [0.0.4] - 2020/10/2
* add Apache License.

## [0.0.5] - 2020/10/2
* github link.

## [0.0.6] - 2020/10/2
* example and documentation added.

## [0.0.7] - 2021/03/24
* null safety

example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_expandable_example/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


test/flutter_expanded_text_widget_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_expanded_text_widget/flutter_expanded_text_widget.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
    expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  });
}


lib/expandable_text_widget.dart
import 'package:flutter/material.dart';

class ExpandableTextWidget extends StatefulWidget {
  const ExpandableTextWidget(
      {Key? key,
      this.text,
      this.title,
      this.margin,
      this.textStyle,
      this.backgroundColor,
      this.downIcon,
      this.upIcon,
      this.shape,
      this.imageChild,
      this.elevation})
      : assert(text != null && title != null),
        super(key: key);

  /// Text, should not be null
  final String? text;

  /// Style of text
  final TextStyle? textStyle;

  /// Typically a [Text] widget that contains a description of the current
  /// contents of the app.
  final Widget? title;

  /// Color of the default background widget
  final Color? backgroundColor;

  /// The z-coordinate at which to place this card. This controls the size of
  /// the shadow below the card.
  ///
  /// Defines the card's [Material.elevation].
  ///
  /// If this property is null then [CardTheme.elevation] of
  /// [ThemeData.cardTheme] is used. If that's null, the default value is 1.0.
  final double? elevation;

  /// The shape of the card's [Material].
  ///
  /// Defines the card's [Material.shape].
  ///
  /// is used. If that's null then the shape will be a [RoundedRectangleBorder]
  /// with a circular corner radius of 4.0.
  final ShapeBorder? shape;

  /// image widget
  final Widget? imageChild;

  /// down icon and widget
  final Widget? downIcon;

  /// up icon and widget
  final Widget? upIcon;

  ///
  final MainAxisAlignment alignmentBottomIcon = MainAxisAlignment.start;

  /// The empty space that surrounds the card.
  ///
  /// Defines the card's outer [Container.margin].
  ///
  /// If this property is null then [CardTheme.margin] of
  /// [ThemeData.cardTheme] is used. If that's null, the default margin is 4.0
  /// logical pixels on all sides: `EdgeInsets.all(4.0)`.
  final EdgeInsetsGeometry? margin;

  @override
  _ExpandableTextState createState() => _ExpandableTextState();
}

class _ExpandableTextState extends State<ExpandableTextWidget> {
  bool descTextShowFlag = false;

  @override
  Widget build(BuildContext context) {
    return Card(
      color: widget.backgroundColor,
      margin: widget.margin,
      elevation: widget.elevation,
      shape: widget.shape,
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        mainAxisAlignment: MainAxisAlignment.center,
        mainAxisSize: MainAxisSize.min,
        children: [
          Container(
            child: widget.imageChild,
          ),
          Container(
            margin: EdgeInsets.all(10.0),
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: <Widget>[
                widget.title!,
                Text(
                  widget.text!,
                  maxLines: descTextShowFlag ? 16 : 2,
                  textAlign: TextAlign.start,
                  style: widget.textStyle,
                ),
                Row(
                  mainAxisAlignment: widget.alignmentBottomIcon,
                  children: <Widget>[
                    if (descTextShowFlag)
                      InkWell(
                        child: widget.upIcon,
                        onTap: () {
                          setState(() => descTextShowFlag = !descTextShowFlag);
                        },
                      )
                    else
                      InkWell(
                        child: widget.downIcon,
                        onTap: () {
                          setState(() => descTextShowFlag = !descTextShowFlag);
                        },
                      )
                  ],
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }
}


lib/flutter_expanded_text_widget.dart
library flutter_expanded_text_widget;

import 'package:flutter/material.dart';

class ExpandableTextWidget extends StatefulWidget {

  const ExpandableTextWidget({
    Key key,
    this.text,
    this.title,
    this.textStyle,
    this.backgroundColor,
    this.downIcon,
    this.upIcon,
    this.shape,
    this.imageChild,
    this.elevation
  }): assert(text != null && title != null),
        super(key: key);

  final String text;
  final TextStyle textStyle;
  final Widget title;
  final Color backgroundColor;
  final double elevation;
  final ShapeBorder shape;
  final Image imageChild;
  final Widget downIcon;
  final Widget upIcon;

  @override
  _ExpandableTextState createState() => _ExpandableTextState();
}

class _ExpandableTextState extends State<ExpandableTextWidget> {
  bool descTextShowFlag = false;

  @override
  Widget build(BuildContext context) {
    return Card(
      color: Colors.white,
      margin: EdgeInsets.all(8),
      elevation: 2.0,
      shape: RoundedRectangleBorder(
        borderRadius: BorderRadius.all(Radius.circular(8)),
      ),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        mainAxisAlignment: MainAxisAlignment.center,
        mainAxisSize: MainAxisSize.min,
        children: [
          Container(
            color: Color(0xff00cccc),
            child: Center(
              child: widget.imageChild,
            ),
          ),
          Container(
            margin: EdgeInsets.all(16.0),
            child: new Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: <Widget>[
                widget.title,
                Text(
                  widget.text,
                  maxLines: descTextShowFlag ? 16 : 2,
                  textAlign: TextAlign.start,
                  style: widget.textStyle,
                ),
                Row(
                  mainAxisAlignment: MainAxisAlignment.start,
                  children: <Widget>[
                    if (descTextShowFlag)
                      IconButton(
                        icon: widget.upIcon,
                        onPressed: () {
                          setState(() {
                            descTextShowFlag = !descTextShowFlag;
                          });
                        },
                      )
                    else
                      IconButton(
                        icon: widget.downIcon,
                        onPressed: () {
                          setState(() {
                            descTextShowFlag = !descTextShowFlag;
                          });
                        },
                      )
                  ],
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }
}


flutter_plugin/android/src/main/java/com/quanticuz/flutter_plugin/FlutterPlugin.java
package com.quanticuz.flutter_plugin;

import androidx.annotation.NonNull;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/** FlutterPlugin */
public class FlutterPlugin implements FlutterPlugin, MethodCallHandler {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private MethodChannel channel;

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
    channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "flutter_plugin");
    channel.setMethodCallHandler(this);
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    if (call.method.equals("getPlatformVersion")) {
      result.success("Android " + android.os.Build.VERSION.RELEASE);
    } else {
      result.notImplemented();
    }
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    channel.setMethodCallHandler(null);
  }
}


example/lib/main.dart
import 'package:expandable_text_widget/expandable_text_widget.dart';
import 'package:flutter/material.dart';
import 'package:flutter_svg/flutter_svg.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Expandable Text Widget Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: ExpandableDemo(),
      debugShowCheckedModeBanner: false,
    );
  }
}

class ExpandableDemo extends StatefulWidget {
  @override
  _ExpandableDemoState createState() => _ExpandableDemoState();
}

class _ExpandableDemoState extends State<ExpandableDemo> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text("Expandable Text Widget Demo"),
      ),
      body: SingleChildScrollView(
        child: Container(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.start,
            crossAxisAlignment: CrossAxisAlignment.stretch,
            children: <Widget>[
              ExpandableTextWidget(
                backgroundColor: Colors.white,
                elevation: 2.0,
                shape: RoundedRectangleBorder(
                  borderRadius: BorderRadius.all(Radius.circular(8.0)),
                ),
                title: Text(
                  "Sarlavha",
                  style: TextStyle(color: Colors.grey[800], fontSize: 24),
                ),
                imageChild: Image.asset(
                  "assets/logo.png",
                  fit: BoxFit.scaleDown,
                ),
                textStyle: TextStyle(color: Colors.black, fontSize: 16),
                text:
                    "The color can \n be set with the splashColor property. \nThe splash size is \ndependent on the size of the child widget passed in - which is constrained by the minRadius and maxRadius parameters.",
                downIcon: SvgPicture.asset(
                  "assets/down.svg",
                  fit: BoxFit.scaleDown,
                ),
                upIcon: SvgPicture.asset(
                  "assets/up.svg",
                  color: Colors.black,
                  fit: BoxFit.scaleDown,
                ),
              ),
              SizedBox(
                height: 20,
              ),
              ExpandableTextWidget(
                backgroundColor: Colors.white,
                elevation: 2.0,
                shape: RoundedRectangleBorder(
                  borderRadius: BorderRadius.all(Radius.circular(8.0)),
                ),
                title: Text(
                  "Sarlavha",
                  style: TextStyle(color: Colors.grey[800], fontSize: 24),
                ),
                textStyle: TextStyle(color: Colors.black, fontSize: 16),
                text:
                    "The color can \n be set with the splashColor property. \nThe splash size is \ndependent on the size of the child widget passed in - which is constrained by the minRadius and maxRadius parameters.",
                downIcon: Text("show"),
                upIcon: Text("hide"),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
